File: DesmosTest/validateClassCodes.js


function validateClassCodes(code, distracting_words, existing_codes) {
    code = code.toLowerCase();
    if (existing_codes.includes(code)) {
        console.log("Code already has been used.");
        return false;
    } else {
        if (containsDistractingWords(code, distracting_words)) {
            console.log("Code contains a distracting word:  ", code);
            return false;
        } else {
            existing_codes.push(code);
            console.log("Adding code to existing codes: ", existing_codes);
            return true;
        }
    }
}

function containsDistractingWords(code, distracting_words) {
    if (foundWordInOrder(code, distracting_words)) {
        console.log("foundWordInOrder returned true!");
        return true;
    } else if (foundAllLettersInCode(code, distracting_words)) {
        console.log("foundAllLettersInCode returned true!");
        return true;
    } else {
        console.log("Code is good!");
        return false;
    }

}

function foundWordInOrder(code, distracting_words) {
    //returns true if the code contains distracting_words
    //in consecutive order (RATS42)
    for (let i = 0; i < distracting_words.length; i++) {
        if (distracting_words[i].toLowerCase().includes(code)) {
            console.log("found entire word in code: ", code);
            return true;
        } else {
            return false;
        }
    }

}

function foundAllLettersInCode(code, distracting_words) {
    //returns true if code contains all letters of
    //any distracting word

    //for each distracting word
    for (let i = 0; i < distracting_words.length; i++) {
        let distracting_word = distracting_words[i].toLowerCase();
        console.log(`Is ${code} in this distracting word: ${distracting_word}`);
        //look at every code character and see if it is in distracting word
        for (let j = 0; j < code.length; j++) {
            idx = distracting_word.indexOf(code[j])
            if (idx >= 0) {
                distracting_word = distracting_word.slice(0, idx) + distracting_word.slice(idx + 1);
                console.log(`     Found letter ${code[j]}, removing from distracting_word: ${distracting_word}`);
            }
            if (distracting_word.length === 0) {
                console.log("     foundAllLettersInCode:  ", distracting_word.length);
                return true;
            }
        }

    }
    //if we made it here, all of the letters of any distracting word is not in the code
    return false;

}

let distracting_words = [
    "darn",
    "rats",
    "egg",
    "fuzzy",
    "kthx",
    "haha",
    "ugh",
    "777",
    "cheese",
]

let existing_codes = [];

let testCodes = ['rats', 'RRATTS', 'TTRRTTTAASS', 'abc', 'def', 'ghi' ]

for (let i = 0; i < testCodes.length; i++) {
    if (!validateClassCodes(testCodes[i], distracting_words, existing_codes)) {
        console.log("Undesirable code ... ", testCodes[i], " ... moving on...");
    }
}

File: DesmosTest/validateClassCodes.py


def validateClassCodes(code, distracting_words, existing_codes):
    code = code.lower()
    if code in existing_codes :
        print("Code already has been used.")
        return False
    else:
        if containsDistractingWords(code, distracting_words):
            print("Code contains a distracting word:  ", code)
            return False
        else :
            existing_codes.append(code)
            print("Adding code to existing codes: ", existing_codes)
            return True

def containsDistractingWords(code, distracting_words):
    if foundWordInOrder(code, distracting_words) :
        print("foundWordInOrder returned true!")
        return True
    elif foundAllLettersInCode(code, distracting_words):
        print("foundAllLettersInCode returned true!")
        return True
    else :
        print("Code is good!")
        return False




def foundWordInOrder(code, distracting_words):
    # returns Trueif the code contains distracting_words
    # in consecutive order (RATS42)
    for distracting_word in distracting_words:
        if distracting_word.lower() in code:
            print("found entire word in code: ", code)
            return True
        else :
            return False


def foundAllLettersInCode(code, distracting_words) :
    #returns Trueif code contains all letters of
    #any distracting word

    #for each distracting word
    for distracting_word in distracting_words:
        distracting_word = distracting_word.lower()
        print(f'Is {code} in this distracting word: {distracting_word}')
        #look at every code character and see if it is in distracting word
        for char in code :
            idx = distracting_word.find(char)
            if (idx >= 0):
               # distracting_word = distracting_word.slice(0, idx) + distracting_word.slice(idx + 1);
                distracting_word = distracting_word[:idx] + distracting_word[idx + 1:]
                print('     Found letter {char}, removing from distracting_word: {distracting_word}');

            if len(distracting_word) == 0:
                print("     foundAllLettersInCode:  ", len(distracting_word));
                return True

    #if we made it here, all of the letters of any distracting word is not in the code
    return False


distracting_words = [
    "darn",
    "rats",
    "egg",
    "fuzzy",
    "kthx",
    "haha",
    "ugh",
    "777",
    "cheese",
]

existing_codes = []

testCodes = ['rats', 'RRATTS', 'TTRRTTTAASS', 'abc', 'def', 'ghi' ]

for code in testCodes :
    if  not validateClassCodes(code, distracting_words, existing_codes):
        print("Undesirable code ... ", code, " ... moving on...")
